privilege/privileges/cache.go

```diff
@@ -263,6 +271,66 @@ func (p *MySQLPrivilege) FindAllRole(activeRoles []*auth.RoleIdentity) []*auth.R
 	return ret
 }
 
+func (p *MySQLPrivilege) IncLoginFail(user, host string) int {
+	key := user + "@" + host
+	cnt, exist := p.PwdErrorCnt[key]
+	if exist {
+		p.PwdErrorCnt[key] = cnt + 1
+		return cnt + 1
+	}
+	p.PwdErrorCnt[key] = 1
+	return 1
+}
+
+func (p *MySQLPrivilege) ClearLoginFail(user, host string) {
+	key := user + "@" + host
+	if p.PwdErrorCnt == nil {
+		p.PwdErrorCnt = make(map[string]int)
+	}
+	p.PwdErrorCnt[key] = 0
+}
+
+func (p *MySQLPrivilege) LockAccount(user, host string, sctx sessionctx.Context) error {
+	lock := types.CurrentTime(0)
+	ctx := context.Background()
+	sql := fmt.Sprintf("insert into mysql.login_blacklist(USER, HOST, lock_time) values('%s', '%s', '%s') on duplicate key update lock_time = '%s'", user, host, lock.String(), lock.String())
+	_, err := sctx.(sqlexec.SQLExecutor).Execute(ctx, sql)
+	return err
+}
+
+func (p *MySQLPrivilege) CheckAccountLock(user, host string, sctx sessionctx.Context, limit time.Duration) bool {
+	recs, exist := p.BlackList[user]
+	if exist {
+		for _, r := range recs {
+			if r.Host == host {
+				t := r.startTime
+				if time.Now().Sub(t) > limit*time.Second {
+					ctx := context.Background()
+					sql := fmt.Sprintf("delete from mysql.login_blacklist where user = '%s' and host = '%s'", user, host)
```

Comment:
maybe we should remove it from p.BlackList to avoid duplicate delete stmt, but this will make p.BlackList also be goroutine-safe :D